### DISCUSSION TITLE
Security and (Continuous?) Source Code Analysis Tools

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#279] [Ideas]
**Basic**:

1. npm-audit will be enabled by default.
2. We will also follow best practices with "Dependabot"
3. https://docs.github.com/en/code-security/supply-chain-security/managing-vulnerabilities-in-your-projects-dependencies/configuring-dependabot-security-updates

**Advanced**:
We will have to research and integrate some kind of SAST tool to check source code of each commit. (We will have to integrate this into some kind of pre-commit hook)

Some resources for research:
https://docs.npmjs.com/auditing-package-dependencies-for-security-vulnerabilities
https://snyk.io/learn/javascript-security/
https://dev.to/bscript/javascript-security-checklist-af
https://www.securecoding.com/blog/most-common-security-vulnerabilities-using-javascript/
https://geekflare.com/nodejs-security-scanner/
https://github.com/topics/security-audit?l=javascript&o=desc&s=
https://owasp.org/www-community/Source_Code_Analysis_Tools
Apparently coverity has free support for Open Source Packages. Need to research this more and see if we can adopt for our project: https://scan.coverity.com/
Open source alternatives to Coverity: https://alternativeto.net/software/coverity-scan/?license=free

Security enhancement will be an ongoing process since we too are learning the best Practices.

How do we check for security of running client as well as server/hosting? We need to get some help from security specialists.

**Comment by kcramakrishna** [comment]
Other articles:
https://blog.sqreen.com/nodejs-security-best-practices/
https://ngninja.com/posts/secure-coding-javascript
https://github.com/validatorjs/validator.js
https://www.javascriptjanuary.com/blog/defensive-javascript
https://blog.bitsrc.io/8-steps-to-secure-javascript-in-2021-6d54d5415264
https://snyk.io/learn/javascript-security/
https://docs.npmjs.com/auditing-package-dependencies-for-security-vulnerabilities
https://docs.npmjs.com/cli/v7/commands/npm-audit
https://www.whitesourcesoftware.com/free-developer-tools/blog/npm-audit/

Linting and ‘Strict Mode’ set to on should make things secure too.

**Reply by kcramakrishna** [reply]
General best Practices in Secure development:
https://dzone.com/articles/10-steps-to-secure-software
1. Validate Input
2. Prevent Injection Attacks
3. Build Security testing into development

Others:
https://snyk.io/learn/secure-coding-practices/
https://www.untitledkingdom.com/blog/how-to-take-care-of-security-in-software-development-software-development-life-cycle
https://www.ptsecurity.com/ww-en/analytics/knowledge-base/how-to-approach-secure-software-development/
https://snyk.io/learn/secure-sdlc/
https://www.synopsys.com/blogs/software-security/infuse-security-into-your-software-development-life-cycle/

**Reply by kcramakrishna** [reply]
We need to incorporate basic penetration testing:
Maybe use Kali Linux or some kind of Open Source penetration testing for the front end as well as the backend.

**Comment by kcramakrishna** [comment]
Also need to figure out the best practices for hiding keys, passwords etc . These should not be in source code or on git.

**Comment by kcramakrishna** [comment]
https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner

**Comment by kcramakrishna** [comment]
SonarQube seems to be the standard. Here are some links:
https://plugins.jetbrains.com/plugin/7973-sonarlint
https://www.sonarsource.com/plans-and-pricing/community/ 

Comparing Synk and SonaQube: (Surprisingly NodeJSScan comes out on top)
https://devsecops-pipelines.ayushpriya.tech/comparing_sast_tools/

**Comment by yureshwar** [comment]
Hi,

I am not sure what you were asking can you explain more on this?

Thanks & Regards,
Yureshwar Ravuri
Mail: ***@***.***

**Comment by kcramakrishna** [comment]
This is a cool tool to scan dockers for vulnerabilities: trivy is a scanner for docker containers. I think its pretty interesting.  
https://github.com/aquasecurity/trivy

**Reply by kcramakrishna** [reply]
One more laternative: https://github.com/anchore/grype 

https://gitlab.com/gitlab-org/gitlab/-/issues/327174

**Comment by yureshwar** [comment]
Implemented eslint based configuration for source code analysis via git hooks. Followed the tutorial from the below given link

https://www.antstack.io/blog/adding-git-hooks-to-your-project/

**Comment by yureshwar** [comment]
There is an good package which can be easily integrated into both NPM based packages and Java packages

https://www.reshiftsecurity.com/

Need to perform an integration with open source one.

**Comment by yureshwar** [comment]
Need to check on

https://snyk.io/product/open-source-security-management/ 
https://www.npmjs.com/package/better-npm-audit

**Comment by kcramakrishna** [comment]
Input validations in node/react
https://www.npmjs.com/package/node-input-validator
https://www.npmjs.com/package/validator
https://www.npmjs.com/package/react-form-input-validation

Other:
https://webcache.googleusercontent.com/search?q=cache:z_UxS6CSYvwJ:https://openbase.com/categories/js/best-nodejs-data-validation-libraries+&cd=8&hl=en&ct=clnk&gl=in&client=firefox-b-d

**Comment by kcramakrishna** [comment]
Data validation in REST and backend:
https://quarkus.io/guides/validation 
https://camel.apache.org/camel-quarkus/2.7.x/reference/extensions/bean-validator.html
https://medium.com/geekculture/how-to-make-input-validation-with-hibernate-eac6a7b20eb0

Also implement state of the art backend API versioning. Ensure that front end includes the correct backend API version. Will this ractice make things easier and more secure? What are best practices here ?

**Comment by kcramakrishna** [comment]
More on defensive programming:
https://javascript.plainenglish.io/writing-defensive-javascript-c6d42940d30f
https://levelup.gitconnected.com/javascript-best-practices-defensive-programming-9592a9329787

Google for this.

**Comment by SHIFTLEFT-INFOSEC-CONSULTANTS-LLP** [comment]
Below is what we plan to implement for Digital Asssistance from Security point of View. 

1. Threat Model Analysis for major changes
2. Manual Pen Test for all prioritized releases. (Using Burp community version & OWASP ZAP)
3. Automated Security Scans for all releases (Tools currently being evaluated). 
4. Incident Handling - Handling vulnerabilities being reported from external sources. 
5. Continuous Monitoring of 3rd Party Libraries or Servers or Infrastructure to track of known vulnerabilities. ( Mend Bolt currently being evaluated).

**Comment by kcramakrishna** [comment]
https://google.github.io/clusterfuzzlite/ 
https://google.github.io/clusterfuzz/

https://www.cidersecurity.io/blog/research/how-to-secure-your-open-source-project-a-quick-guide-for-developers/

We need to be able to publish security reports for every public release: Something like below
https://www.cvedetails.com/vulnerability-list/vendor_id-45/product_id-887/year-2022/Apache-Tomcat.html
https://tomcat.apache.org/security-9.html

**Comment by kcramakrishna** [comment]
https://github.com/OWASP/wstg/tree/master/document
https://owasp.org/www-project-samm/
owasp zap

**Comment by kcramakrishna** [comment]
https://www.helpnetsecurity.com/2024/09/10/open-source-cybersec-tools/ some of these tools are interesting.

**Comment by kcramakrishna** [comment]
https://www.darkrelay.com/post/owasp-api-top-10-testing-guide

**Comment by kcramakrishna** [comment]
Short notes on all security steps planned for our application: (Need to create issues for each of these and close them.)
1. Server and Docker Security Configurations: mariaDB, elastic, quarkus, NEST, React, PostGres
2. input sanitisation and validations: NEST, Quarkus
3. Network communications - SSL/TLS between every docker and service
4. Ingress: ngnix (+modSecurity)
5. Misc: Rate-limiters in ngnix AND in frameworks of NEST, quarkus etc. 
6. Applications:6.a: MariaDB: hardening and user restriction6.b: Elastic security hardening6.c: PostGRessecurity Hardening6.d: KeyCLoak security Hardening6.e: any other servers ?  
7. Keycloak - Why do we need to restart to make the token changes effective ?
8. Need to configure ngnix+ModSecurity
